File: backend/src/services/excelService.ts
import ExcelJS from 'exceljs';
import { ReportData } from '../types';
import { TIME_SLOTS } from '../constants';
import { Buffer } from 'buffer';
import { logger } from '../utils/logger';

export class ExcelService {
  async generateExcel(data: ReportData, templateBuffer?: Buffer): Promise<Buffer> {
    const workbook = new ExcelJS.Workbook();
    let worksheet: ExcelJS.Worksheet;

    try {
      if (templateBuffer) {
        try {
          await workbook.xlsx.load(templateBuffer);
          worksheet = workbook.getWorksheet(1) || workbook.addWorksheet('EOD Report');
        } catch (error) {
          logger.warn('Failed to load template, using default layout');
          worksheet = workbook.addWorksheet('EOD Report');
          this.setupDefaultLayout(worksheet, data);
        }
      } else {
        worksheet = workbook.addWorksheet('EOD Report');
        this.setupDefaultLayout(worksheet, data);
      }

      this.fillReportData(worksheet, data);
      this.applyFormatting(worksheet);

      const buffer = await workbook.xlsx.writeBuffer();
      return Buffer.from(buffer);
    } catch (error) {
      logger.error('Excel generation error:', error);
      throw new Error('Failed to generate Excel file');
    }
  }

  private setupDefaultLayout(worksheet: ExcelJS.Worksheet, data: ReportData): void {
    // Title
    worksheet.mergeCells('B1:C1');
    const titleCell = worksheet.getCell('B1');
    titleCell.value = 'EOD REPORT';
    titleCell.font = { name: 'Calibri', size: 16, bold: true };
    titleCell.alignment = { horizontal: 'center', vertical: 'center' };

    // Employee details
    const details = [
      ['B3', 'Name Of Employee', 'C3'],
      ['B4', 'Position', 'C4'],
      ['B5', 'Date', 'C5']
    ];

    details.forEach(([labelCell, label, valueCell]) => {
      worksheet.getCell(labelCell).value = label;
      worksheet.getCell(labelCell).font = { bold: true };
    });

    // Headers
    const headers = ['B7', 'C7', 'D7'];
    const headerLabels = ['Time', 'Activity', 'Description'];
    headers.forEach((cell, index) => {
      worksheet.getCell(cell).value = headerLabels[index];
      worksheet.getCell(cell).font = { bold: true };
    });

    // Set column widths
    worksheet.getColumn('B').width = 18;
    worksheet.getColumn('C').width = 28;
    worksheet.getColumn('D').width = 35;

    // Page setup
    worksheet.pageSetup = {
      orientation: 'portrait',
      fitToPage: true,
      fitToWidth: 1,
      fitToHeight: 0
    };
  }

  private fillReportData(worksheet: ExcelJS.Worksheet, data: ReportData): void {
    worksheet.getCell('C3').value = data.employee_name;
    worksheet.getCell('C4').value = data.position;
    worksheet.getCell('C5').value = data.date;

    const startRow = 8;
    data.schedule.forEach((entry, index) => {
      const row = startRow + index;
      worksheet.getCell(`B${row}`).value = entry.slot;
      worksheet.getCell(`C${row}`).value = entry.activity;
      worksheet.getCell(`D${row}`).value = entry.description;
    });
  }

  private applyFormatting(worksheet: ExcelJS.Worksheet): void {
    const thinBorder = {
      top: { style: 'thin' as const },
      left: { style: 'thin' as const },
      bottom: { style: 'thin' as const },
      right: { style: 'thin' as const }
    };

    // Apply borders to schedule
    const startRow = 7;
    const endRow = 15;
    for (let row = startRow; row <= endRow; row++) {
      for (let col = 2; col <= 4; col++) {
        const cell = worksheet.getCell(row, col);
        cell.border = thinBorder;
        cell.alignment = { 
          vertical: 'middle',
          wrapText: true
        };
      }
    }

    // Center headers
    const headerRow = worksheet.getRow(7);
    headerRow.alignment = { horizontal: 'center', vertical: 'middle' };
    headerRow.height = 25;

    // Add some styling to employee info
    ['C3', 'C4', 'C5'].forEach(cell => {
      const cellObj = worksheet.getCell(cell);
      cellObj.border = thinBorder;
      cellObj.alignment = { vertical: 'middle' };
    });
  }
}

File: backend/app/services/excel_service.py
from io import BytesIO
from openpyxl import load_workbook, Workbook
from openpyxl.styles import Font, Alignment, Border, Side
from datetime import datetime
import os

# Default slots (used if no time_slots provided)
DEFAULT_TIME_SLOTS = [
    "10:00 am to 11:00 am",
    "11:00 am to 12:00 pm",
    "12:00 pm to 1:00 pm",
    "1:00 pm to 2:00 pm",
    "2:00 pm to 3:00 pm",
    "3:00 pm to 4:00 pm",
    "4:00 pm to 5:00 pm",
    "5:00 pm to 6:00 pm"
]

# Helper: Format date to DD-MM-YYYY
def format_date_for_excel(date_val):
    if not date_val:
        return ""
    if isinstance(date_val, str):
        try:
            dt = datetime.strptime(date_val, "%Y-%m-%d")
            return dt.strftime("%d-%m-%Y")
        except ValueError:
            return date_val
    elif isinstance(date_val, datetime):
        return date_val.strftime("%d-%m-%Y")
    return str(date_val)

def create_excel_bytes(schedule_data, template_bytes=None, time_slots=None):
    if time_slots is None:
        time_slots = DEFAULT_TIME_SLOTS

    wb = None
    ws = None

    # 1. Try to load template
    if template_bytes is None:
        template_path = os.path.join(os.path.dirname(__file__), '..', '..', 'EOD_SAMPLE_FINAL.xlsx')
        if os.path.exists(template_path):
            with open(template_path, 'rb') as f:
                template_bytes = f.read()
            print(f"✅ Loaded template from {template_path}")

    if template_bytes:
        try:
            wb = load_workbook(BytesIO(template_bytes))
            ws = wb.active
            print("✅ Template loaded successfully")
        except Exception as e:
            print(f"⚠️ Failed to load template: {e} – using built-in")
            wb = None

    # 2. Built-in layout if no template
    if wb is None:
        wb = Workbook()
        ws = wb.active
        ws.title = "EOD Report"

        ws.merge_cells('B1:C1')
        cell = ws['B1']
        cell.value = "EOD REPORT"
        cell.font = Font(name='Calibri', size=16, bold=True)
        cell.alignment = Alignment(horizontal='center', vertical='center')

        ws['B3'] = "Name Of Employee"
        ws['C3'] = ""
        ws['B4'] = "Position"
        ws['C4'] = ""
        ws['B5'] = "Date"
        ws['C5'] = ""

        for r in [3,4,5]:
            ws[f'B{r}'].font = Font(bold=True)

        ws['B7'] = "Time"
        ws['C7'] = "Activity"
        ws['D7'] = "Description"
        for col in ['B','C','D']:
            ws[f'{col}7'].font = Font(bold=True)

        ws.column_dimensions['B'].width = 18
        ws.column_dimensions['C'].width = 28
        ws.column_dimensions['D'].width = 35

        thin = Border(left=Side(style='thin'), right=Side(style='thin'),
                      top=Side(style='thin'), bottom=Side(style='thin'))
        for row in ws.iter_rows(min_row=7, max_row=len(time_slots)+7, min_col=2, max_col=4):
            for cell in row:
                cell.border = thin
        for cell_addr in ['C3','C4','C5']:
            ws[cell_addr].border = thin

    # 3. Fill data with formatted date
    try:
        ws['C3'] = schedule_data.get("employee_name", "")
        ws['C4'] = schedule_data.get("position", "")
        ws['C5'] = format_date_for_excel(schedule_data.get("date", ""))
    except Exception as e:
        print(f"⚠️ Could not fill employee details: {e}")

    # 4. Fill schedule rows
    start_row = 8
    for idx, slot in enumerate(time_slots):
        row = start_row + idx
        entry = next((item for item in schedule_data.get("schedule", []) if item.get("slot") == slot), None)
        activity = entry.get("activity", "") if entry else ""
        description = entry.get("description", "") if entry else ""

        if "Lunch Break" in slot or "lunch" in slot.lower():
            activity = "Lunch Break"
            description = "Lunch Break"

        try:
            ws[f'C{row}'] = activity
            ws[f'D{row}'] = description
        except Exception as e:
            print(f"⚠️ Could not fill row {row}: {e}")

    output = BytesIO()
    wb.save(output)
    output.seek(0)
    return output.getvalue()